consensus/src/detail/mod.rs

```diff
@@ -200,6 +208,119 @@ impl Consensus {
         let block = consensus_ref.db_tx.get_block(id).map_err(BlockError::from)?;
         Ok(block)
     }
+
+    pub fn get_block_index(&self, id: &Id<Block>) -> Result<Option<BlockIndex>, BlockError> {
+        self.make_ro_db_tx().db_tx.get_block_index(id).map_err(BlockError::from)
+    }
+
+    pub fn get_header_from_height(
+        &self,
+        height: &BlockHeight,
+    ) -> Result<Option<BlockHeader>, BlockError> {
+        let consensus_ref = self.make_ro_db_tx();
+        let id = consensus_ref
+            .db_tx
+            .get_block_id_by_height(height)?
+            .ok_or(BlockError::NotFound)?;
+        consensus_ref.db_tx.get_block_header_by_id(&id).map_err(BlockError::from)
+    }
+
+    pub fn get_locator(&self) -> Result<Vec<BlockHeader>, BlockError> {
+        let id = self.get_best_block_id()?.ok_or(BlockError::NotFound)?;
+        let height = self.get_block_height_in_main_chain(&id)?.ok_or(BlockError::NotFound)?;
+
+        let headers = itertools::iterate(0, |&i| if i == 0 { 1 } else { i * 2 })
+            .take_while(|i| (height - BlockDistance::new(*i)).is_some())
+            .map(|i| {
+                self.get_header_from_height(
+                    &(height - BlockDistance::new(i)).expect("distance to be valid"),
+                )
+            })
+            .filter(|res| res.is_ok())
+            .filter_map(|res| res.unwrap())
+            .collect::<Vec<_>>();
+
+        Ok(headers)
+    }
+
+    // TODO: use `ConsensusRef::is_block_in_main_chain()` implementation instead, how?
+    fn is_block_in_main_chain(&self, block_index: &BlockIndex) -> Result<bool, BlockError> {
+        let consensus_ref = self.make_ro_db_tx();
+        let height = block_index.get_block_height();
+        let id_at_height =
+            consensus_ref.db_tx.get_block_id_by_height(&height).map_err(BlockError::from)?;
+        match id_at_height {
+            Some(id) => Ok(id == *block_index.get_block_id()),
+            None => Ok(false),
+        }
+    }
+
+    pub fn get_headers(&self, locator: Vec<BlockHeader>) -> Result<Vec<BlockHeader>, BlockError> {
+        // use genesis block if no common ancestor with better block height is found
+        let mut best = BlockHeight::new(0);
+
+        for header in locator.iter() {
```

Comment:
Based on our discussion yesterday it was decided that this is not needed but I'd ask you to take a look again. The `get_ancestor()` function uses block height to find the ancestor block index (meaning the height must be known) whereas this function tries to find the height of the block index so it can collect the correct headers from that point on until it reaches the tip. I don't see a straight-forward way of using the `get_ancestor()` function to advantage here.

Reply:
We should have a function here that gets the common ancestor. I seem to have confused these two. Unfortunately, this function doesn't exist right now. Put a TODO to change this function in the future to use the `last_common_ancestor()` implementation when it exists.

Reply:
I unresolved this because we have an update. Roy has implemented `last_common_ancestor()`. Please use it. No need for a TODO.